asterius/src/Asterius/Types/SymbolMap.hs

```diff
@@ -0,0 +1,239 @@
+-- of the unique of their 'EntitySymbol'-key.
+elems :: SymbolMap a -> [a]
+elems (SymbolMap m) = map snd $ IM.elems m
+
+-- | /O(min(n,W))/. Lookup the value at an 'EntitySymbol' in the map.
+lookup :: EntitySymbol -> SymbolMap a -> Maybe a
+lookup k (SymbolMap m) = snd <$> IM.lookup (getKey $ getUnique k) m
+
+-- | /O(min(n,W))/. Find the value at an 'EntitySymbol'. Calls 'error' when the
+-- element can not be found.
+(!) :: HasCallStack => SymbolMap a -> EntitySymbol -> a
+(!) m k = case lookup k m of
+  Just e -> e
+  Nothing -> error "SymbolMap.!: given key is not an element in the map"
+
+infixl 9 !
+
+-- | /O(n)/. Filter all keys/values that satisfy some predicate. NOTE: since we
+-- use 'Key' for indexing and not 'EntitySymbol', the filtering happens on the
+-- elements alone (which contain the corresponding 'EntitySymbol').
+filterWithKey :: (EntitySymbol -> a -> Bool) -> SymbolMap a -> SymbolMap a
+filterWithKey p (SymbolMap m) = SymbolMap $ IM.filter (uncurry p) m
+
+-- | The restriction of a map to the keys in a set.
+restrictKeys :: SymbolMap a -> Set.Set EntitySymbol -> SymbolMap a
+restrictKeys (SymbolMap m) s =
+  SymbolMap $ IM.restrictKeys m (IS.fromList $ map (getKey . getUnique) $ Set.toList s)
+
+-- | /O(n)/. Map a function over all values in the map.
+mapWithKey :: (EntitySymbol -> a -> b) -> SymbolMap a -> SymbolMap b
+mapWithKey fn (SymbolMap m) =
+  SymbolMap $ IM.mapWithKey (\_ (k, e) -> (k, fn k e)) m
+
+-- | /O(min(n,W))/. Insert a new key/value pair in the map. If the key is
+-- already present in the map, the associated value is replaced with the
+-- supplied value.
+insert :: EntitySymbol -> a -> SymbolMap a -> SymbolMap a
+insert k e (SymbolMap m) =
+  SymbolMap $ IM.insert (getKey $ getUnique k) (k, e) m
+
+-- | /O(n)/. The set of all keys of the map. NOTE: This function utilizes the
+-- 'Ord' instance for 'EntitySymbol' (because it calls 'Set.fromList'
+-- internally).
+keysSet :: SymbolMap a -> Set.Set EntitySymbol
+keysSet (SymbolMap m) = Set.fromList $ map fst $ IM.elems m
+
+-- | /O(n)/. Return all 'EntitySymbol' keys of the map, in ascending order of
+-- the key of their unique.
+keys :: SymbolMap a -> [EntitySymbol]
+keys (SymbolMap m) = map fst $ IM.elems m
+
+-- | /O(n)/. Thread an accumulating argument through the map, in ascending
+-- order of keys of uniques on the 'EntitySymbol' keys.
+mapAccum :: (a -> b -> (a, c)) -> a -> SymbolMap b -> (a, SymbolMap c)
+mapAccum f z (SymbolMap m) =
+  SymbolMap <$> IM.mapAccum (\a (k, e) -> fmap (k,) $ f a e) z m
+
+-- | /O(n)/. Fold the keys and values in the map using the given
+-- right-associative binary operator. This is a strict variant: each
+-- application of the operator is evaluated before using the result in the next
+-- application. This function is strict in the starting value.
+foldrWithKey' :: (EntitySymbol -> a -> b -> b) -> b -> SymbolMap a -> b
+foldrWithKey' f z (SymbolMap m) = IM.foldrWithKey (\_ (k, e) b -> f k e b) z m
```

Comment:
We should be using `IM.foldrWithKey'` here.

Reply:
Ah, good catch! Fixed :slightly_smiling_face: